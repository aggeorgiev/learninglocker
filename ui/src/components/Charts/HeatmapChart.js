import React from 'react';
import { AutoSizer } from 'react-virtualized';
import { compose } from 'recompose';
import NoData from 'ui/components/Graphs/NoData';
import { wrapLabel } from 'ui/utils/defaultTitles';
import {
  getResultsData,
  getShortModel,
  hasData,
  hiddenSeriesState
} from './Chart';
import {
  Chart as StyledChart,
  BarContainer,
  XAxis as StyledXAxis,
  YAxis as StyledYAxis,
  XAxisLabel,
  ChartWrapper
} from './styled';

// Utility functions
export const getColorIntensity = (value, minValue, maxValue) => {
  const normalized = (value - minValue) / (maxValue - minValue);
  return Math.floor(normalized * 255);
};

export const getHeatmapColor = (value, minValue, maxValue, baseColor = [0, 0, 255]) => {
  const intensity = getColorIntensity(value, minValue, maxValue);
  return `rgb(${255 - intensity}, ${255 - intensity}, 255)`;
};

export const findMinMax = data => {
  let min = Infinity;
  let max = -Infinity;

  // Handle Immutable.js data
  if (data && typeof data.toJS === 'function') {
    data = data.toJS();
  }

  // Handle object format
  if (data && !Array.isArray(data) && typeof data === 'object') {
    data = Object.values(data);
  }

  // Ensure we have an array to work with
  if (!Array.isArray(data)) {
    return [0, 0]; // Return default values if data is invalid
  }

  data.forEach(row => {
    if (typeof row === 'object' && row !== null) {
      Object.entries(row).forEach(([key, value]) => {
        // Handle numeric strings by converting them
        const numValue = Number(value);
        
        // Only process if it's a Series key or value property
        if ((key.startsWith('Series') || key === 'value') && 
            !isNaN(numValue) && 
            value !== null && 
            value !== undefined && 
            value !== '') {
          min = Math.min(min, numValue);
          max = Math.max(max, numValue);
        }
      });
    }
  });

  // Handle case where no valid numbers were found
  if (min === Infinity || max === -Infinity) {
    return [0, 0];
  }

  return [min, max];
};

const DEFAULT_AXES_LABELS = {
  xLabel: 'X-Axis',
  yLabel: 'Y-Axis'
};

const renderHeatmap = (data, labels, model) => ({ width, height }) => {
  const cellWidth = width / (labels.size + 1);
  const cellHeight = height / data.size;
  const [minValue, maxValue] = findMinMax(data.toJS());

  return (
    <div className="w-full h-full relative">
      {/* Y-Axis Labels */}
      <div className="absolute left-0 top-0 w-24">
        {data.map((row, i) => (
          <div
            key={`y-${i}`}
            className="text-sm truncate"
            style={{
              height: `${cellHeight}px`,
              lineHeight: `${cellHeight}px`
            }}
          >
            {getShortModel(data)(row.get('cellId'))}
          </div>
        ))}
      </div>

      {/* Heatmap Grid */}
      <div className="ml-24">
        <div className="flex mb-2">
          {labels.map((label, i) => (
            <div
              key={`x-${i}`}
              className="text-sm text-center truncate"
              style={{ width: `${cellWidth}px` }}
            >
              {label}
            </div>
          ))}
        </div>

        {data.map((row, rowIndex) => (
          <div key={`row-${rowIndex}`} className="flex">
            {labels.map((_, colIndex) => {
              const value = row.get(`Series ${colIndex + 1}`, 0);
              return (
                <div
                  key={`cell-${rowIndex}-${colIndex}`}
                  className="border border-gray-200"
                  style={{
                    width: `${cellWidth}px`,
                    height: `${cellHeight}px`,
                    backgroundColor: getHeatmapColor(value, minValue, maxValue),
                  }}
                >
                  <div className="w-full h-full flex items-center justify-center text-xs">
                    {value}
                  </div>
                </div>
              );
            })}
          </div>
        ))}
      </div>
    </div>
  );
};

const getAxisLabel = (axesLabels, key, model, defaultKey, fallback) => {
  if (axesLabels && axesLabels[key]) {
    return axesLabels[key];
  }
  if (model && model.getIn) {
    return model.getIn([defaultKey, 'searchString']) || fallback;
  }
  return fallback;
};

const renderChart = (component, axesLabels = {}, chartWrapperFn, model) => (
  <StyledChart>
    <BarContainer>
      <StyledYAxis>
      {wrapLabel(getAxisLabel(axesLabels, 'yLabel', model, 'axesvalue', DEFAULT_AXES_LABELS.yLabel))}
      </StyledYAxis>
      <ChartWrapper>
        {chartWrapperFn(component)}
      </ChartWrapper>
    </BarContainer>
    <XAxisLabel>
      <StyledXAxis>
       {wrapLabel(getAxisLabel(axesLabels, 'xLabel', model, 'axesgroup', DEFAULT_AXES_LABELS.xLabel))}
      </StyledXAxis>
    </XAxisLabel>
  </StyledChart>
);

const sortData = data => data.sortBy(e => e.get('id'));

const getSortedData = results => labels =>
  sortData(getResultsData(results)(labels));

const renderChartResults = (labels, data) => 
  renderHeatmap(data, labels);

const renderResults = results => labels => axesLabels => chartWrapperFn => model =>
  renderChart(
    renderChartResults(labels, getSortedData(results)(labels)),
    axesLabels,
    chartWrapperFn,
    model
  );

const HeatmapChart = compose(hiddenSeriesState)((
  {
    results,
    labels,
    axesLabels = DEFAULT_AXES_LABELS,
    chartWrapperFn = component => (<AutoSizer>{component}</AutoSizer>),
    model
  }) => (
    hasData(results)
      ? renderResults(results)(labels)(axesLabels)(chartWrapperFn)(model)
      : <NoData />
  )
);

// Export both the component and testing utilities
export default HeatmapChart;

// Export testing utilities
export const testing = {
  getColorIntensity,
  getHeatmapColor,
  findMinMax
};
