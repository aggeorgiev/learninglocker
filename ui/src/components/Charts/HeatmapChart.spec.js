import React from 'react';
import 'jest-styled-components';
import renderer from 'react-test-renderer';
import { List, fromJS } from 'immutable';
import { Provider } from 'react-redux';
import { createStore } from 'redux';
import HeatmapChart, { getColorIntensity, getHeatmapColor, findMinMax } from './HeatmapChart';

// ... previous tests remain the same ...

describe('Heatmap utility functions', () => {
  test('getColorIntensity should return correct intensity values', () => {
    expect(getColorIntensity(0, 0, 100)).toBe(0);
    expect(getColorIntensity(50, 0, 100)).toBe(127);
    expect(getColorIntensity(100, 0, 100)).toBe(255);
  });

  test('getHeatmapColor should return correct RGB values', () => {
    expect(getHeatmapColor(0, 0, 100)).toBe('rgb(255, 255, 255)');
    expect(getHeatmapColor(50, 0, 100)).toBe('rgb(128, 128, 255)');
    expect(getHeatmapColor(100, 0, 100)).toBe('rgb(0, 0, 255)');
  });

  describe('findMinMax', () => {
    test('should handle Series format data', () => {
      const data = [
        { Series1: 10, Series2: 20 },
        { Series1: 5, Series2: 30 }
      ];
      const [min, max] = findMinMax(data);
      expect(min).toBe(5);
      expect(max).toBe(30);
    });

    test('should handle Immutable.js data', () => {
      const data = fromJS([
        { Series1: 10, Series2: 20 },
        { Series1: 5, Series2: 30 }
      ]);
      const [min, max] = findMinMax(data);
      expect(min).toBe(5);
      expect(max).toBe(30);
    });

    test('should handle object format data', () => {
      const data = {
        row1: { value: 10, other: 'text' },
        row2: { value: 5, other: 'text' },
        row3: { value: 30, other: 'text' }
      };
      const [min, max] = findMinMax(data);
      expect(min).toBe(5);
      expect(max).toBe(30);
    });

    test('should ignore non-numeric values', () => {
      const data = [
        { Series1: '10', Series2: 'invalid', Series3: 20 },
        { Series1: 5, Series2: null, Series3: undefined }
      ];
      const [min, max] = findMinMax(data);
      expect(min).toBe(5);
      expect(max).toBe(20);
    });
  });
});
