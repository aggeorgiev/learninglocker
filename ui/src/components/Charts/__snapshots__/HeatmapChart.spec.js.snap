// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeatmapChart should apply correct color intensity based on values 1`] = `
.c0 {
  height: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c0 svg g text {
  font-size: 12px;
}

.c3 {
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
}

.c5 {
  display: inline-block;
  white-space: nowrap;
}

.c4 {
  text-align: right;
  word-break: break-word;
  overflow: hidden;
  text-overflow: ellipsis;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  position: static;
  margin-right: 15px;
  margin-bottom: 10px;
}

.c2 {
  display: inline-block;
  -webkit-writing-mode: vertical-rl;
  -ms-writing-mode: tb-rl;
  writing-mode: vertical-rl;
  -webkit-transform: rotate(-180deg);
  -ms-transform: rotate(-180deg);
  transform: rotate(-180deg);
  margin: auto 0;
  white-space: pre-wrap;
}

<div
  className="c0"
>
  <div
    className="c1"
  >
    <div
      className="c2"
    >
      Range
    </div>
    <div
      className="c3"
    >
      <div />
    </div>
  </div>
  <div
    className="c4"
  >
    <span
      className="c5"
    >
      Intensity
    </span>
  </div>
</div>
`;

exports[`HeatmapChart should render no data 1`] = `
<h4
  style={
    Object {
      "marginTop": 20,
      "textAlign": "center",
      "width": "100%",
    }
  }
>
  No data in this time range.
</h4>
`;

exports[`HeatmapChart should render with custom axes labels 1`] = `
.c0 {
  height: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c0 svg g text {
  font-size: 12px;
}

.c3 {
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
}

.c5 {
  display: inline-block;
  white-space: nowrap;
}

.c4 {
  text-align: right;
  word-break: break-word;
  overflow: hidden;
  text-overflow: ellipsis;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  position: static;
  margin-right: 15px;
  margin-bottom: 10px;
}

.c2 {
  display: inline-block;
  -webkit-writing-mode: vertical-rl;
  -ms-writing-mode: tb-rl;
  writing-mode: vertical-rl;
  -webkit-transform: rotate(-180deg);
  -ms-transform: rotate(-180deg);
  transform: rotate(-180deg);
  margin: auto 0;
  white-space: pre-wrap;
}

<div
  className="c0"
>
  <div
    className="c1"
  >
    <div
      className="c2"
    >
      THE-Y
    </div>
    <div
      className="c3"
    >
      <div />
    </div>
  </div>
  <div
    className="c4"
  >
    <span
      className="c5"
    >
      THE-X
    </span>
  </div>
</div>
`;

exports[`HeatmapChart should render with data 1`] = `
.c0 {
  height: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c0 svg g text {
  font-size: 12px;
}

.c3 {
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
}

.c5 {
  display: inline-block;
  white-space: nowrap;
}

.c4 {
  text-align: right;
  word-break: break-word;
  overflow: hidden;
  text-overflow: ellipsis;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  position: static;
  margin-right: 15px;
  margin-bottom: 10px;
}

.c2 {
  display: inline-block;
  -webkit-writing-mode: vertical-rl;
  -ms-writing-mode: tb-rl;
  writing-mode: vertical-rl;
  -webkit-transform: rotate(-180deg);
  -ms-transform: rotate(-180deg);
  transform: rotate(-180deg);
  margin: auto 0;
  white-space: pre-wrap;
}

<div
  className="c0"
>
  <div
    className="c1"
  >
    <div
      className="c2"
    >
      Categories
    </div>
    <div
      className="c3"
    >
      <div />
    </div>
  </div>
  <div
    className="c4"
  >
    <span
      className="c5"
    >
      Time Periods
    </span>
  </div>
</div>
`;

exports[`HeatmapChart should render with store and model 1`] = `
.c0 {
  height: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c0 svg g text {
  font-size: 12px;
}

.c3 {
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
}

.c5 {
  display: inline-block;
  white-space: nowrap;
}

.c4 {
  text-align: right;
  word-break: break-word;
  overflow: hidden;
  text-overflow: ellipsis;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  position: static;
  margin-right: 15px;
  margin-bottom: 10px;
}

.c2 {
  display: inline-block;
  -webkit-writing-mode: vertical-rl;
  -ms-writing-mode: tb-rl;
  writing-mode: vertical-rl;
  -webkit-transform: rotate(-180deg);
  -ms-transform: rotate(-180deg);
  transform: rotate(-180deg);
  margin: auto 0;
  white-space: pre-wrap;
}

<div
  className="c0"
>
  <div
    className="c1"
  >
    <div
      className="c2"
    >
      Y-Axis
    </div>
    <div
      className="c3"
    >
      <div />
    </div>
  </div>
  <div
    className="c4"
  >
    <span
      className="c5"
    >
      X-Axis
    </span>
  </div>
</div>
`;
